### DISCUSSION TITLE
MAP:.5-.95 for box detection seems different from MAP:.5-.95 from torchmetrics

### DISCUSSION THREAD

**Discussion by lhericourt** [ultralytics/ultralytics#22190] [Q&A] [answered]
Hi,

I am running a test on finetuning Yolo11 with ultralytics, and during the training phase at the end of each epoch I store the MAP:.5-.95 from the validation set in mlflow. 

A the end of the training, I compute for the best model the same MAP:.5-.95 but this time with the metric from torchmetrics on the train, validation and test set
=> And **I get a significant difference** (14%) on the validation set between the MAP:.5-.95 **from ultralytics** and the MAP:.5-.95 from **torchmetrics**.

After going through the code of Ultralytics and torchmetrics I came to that conclusion that the difference is how both metric are computed:
    - In the curve precision / recall, if recall 1 is never reached:
        + Torchmetric: set directly the precision to 0
        + Ultralytics does not, they interpolate between the highest recall and 1

Are you aware of this difference ? What is the metric usually used is common benchmarks, the one from ultralytics or the one from torchmetrics ?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @lhericourt, thanks for opening this thoughtful discussion and for testing YOLO11 with Ultralytics 🚀 This is an automated response — an Ultralytics engineer will follow up here soon to help further.

We recommend a quick visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, and where many common questions are already answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) to help us investigate the metric discrepancy.

To help us diagnose the MAP:.5-.95 differences you’re seeing between Ultralytics and torchmetrics, please include:
- Exact versions: `ultralytics`, `torchmetrics`, `torch`, `python`, CUDA/CUDNN
- Your training/evaluation commands or code snippets for both Ultralytics and torchmetrics
- The precise settings used for evaluation (conf/iou thresholds, NMS configuration, class-agnostic vs per-class, image sizes, augmentation during eval if any)
- Dataset details (format, number of classes) and a few example images/labels if possible
- Logs or the saved validation metrics per epoch (the values stored to MLflow)
- How you select and load the “best model” for your post-training evaluation
- Confirmation that both evaluations use the same data split and preprocessing

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). 🙏

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Please upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify this isn’t already addressed in the newest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- Amazon Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- Docker Image. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment, marked as answer]
This is discussed in https://github.com/ultralytics/ultralytics/issues/17270

**Reply by lhericourt** [reply]
OK thanks a lot for the answer, indeed I did not see this discussion.
Do you have any idea when the fix will be merged ?

**Reply by glenn-jocher** [reply]
Thanks for following up — we can’t provide timelines for merges, but we’ll update the thread once it lands; for context, Ultralytics computes AP using COCO’s 101‑point interpolation with a precision envelope and recall sentinels (see the compute_ap reference), which is what most benchmarks follow; if you can share a minimal repro showing the 14% gap on the same predictions/IoU settings, we’ll validate it during review. [compute_ap reference](https://docs.ultralytics.com/reference/utils/metrics/#ultralytics.utils.metrics.compute_ap)

**Reply by lhericourt** [reply]
Hi, I am sorry but to reproduce this I would need to provide you the data that are sensitive.

**Reply by glenn-jocher** [reply]
Totally understood—no need to share raw data. If you can either reproduce the gap on a public split (e.g., COCO8) with your torchmetrics script or share only the per-class PR arrays we use (no images/filenames), we can verify the interpolation difference; you can export them like this and attach the resulting NPZ, plus the text output of `yolo checks` for versions:

```python
from ultralytics import YOLO
import numpy as np

res = YOLO("best.pt").val(data="data.yaml", plots=False)
np.savez("pr_curves.npz", p=res.box.p_curve, r=res.box.r_curve, x=res.box.px)
```

For context on what these arrays represent in our evaluator, see the AP/PR curve internals in the [ap_per_class reference](https://docs.ultralytics.com/reference/utils/metrics/#ultralytics.utils.metrics.ap_per_class).

**Reply by lhericourt** [reply]
Hi, I could not be able to reproduce the 14% gap, however usually reproduce a gap around 5%.
Simple use case to reproduce: training Yolov11-n to detect the only class 'drawer', training on 500 images with drawers from ADE20k dataset for 500 epochs. I got a 6% gap on validation.

**Comment by JohnLYYN123** [comment]
Hello,

I hope this message finds you all well.

Just a small follow-up to the previous discussions. As mentioned in the thread, the observed ~14% difference between the two implementations is quite significant, and such a discrepancy could affect the interpretation and application of the model across different domains. 

since [PR #19738](https://github.com/ultralytics/ultralytics/pull/19738) has not been merged yet, could you please clarify which mAP implementation should be used for quantitative reporting and analysis?

Thank you!

**Reply by glenn-jocher** [reply]
Good question—until that PR is reviewed and merged, please report the COCO‑style mAP50–95 from `model.val()` (i.e., `results.box.map`) for comparability across papers and benchmarks; if you also share TorchMetrics results, label them explicitly as a different definition and present both. For context on what we report and how to read it, see the mAP overview in [Mean Average Precision (mAP) Explained](https://www.ultralytics.com/glossary/mean-average-precision-map) and our [Performance Metrics Deep Dive](https://docs.ultralytics.com/guides/yolo-performance-metrics/).

**Reply by JohnLYYN123** [reply]
Hello, 

Thank you so much for the detailed explanation! I really appreciate it.

But I’m still a bit confused. If the current implementations for calculating AP and mAP are already correct, what was the reason behind the PR suggesting potential fixes?

I am now a bit concerned with validity of the numbers reported when using the validation mode. 

Thank you very much for your clarification.

**Reply by Y-T-G** [reply]
You should use mAP from COCOEval for reporting

**Reply by JohnLYYN123** [reply]
Thank you again for the quick reply.

I just want to double confirm since I see a small discrepancy between the most recent two answers. 

For official reporting, should I rely on the evaluation metrics (like mAP) reported by Ultralytics, or should I instead use pycocotools’ COCOeval results?

Thank you again for your time and clarification!

**Reply by Y-T-G** [reply]
COCOEval
